### PR TITLE
LocationCache ConcurrentDict lock contention fix

### LINKED ISSUE #553: Lock contention in LocationCache
LocationCache is suing ConcurrentDictionary.Count, which used a blocking lock underneath.

### PR DESCRIPTION
**Pull request by kirankumarkolli**
## Description

LocationCache used ConcurrentDictionay.Count which serializes through lock resulting in high resource contention. 

Usage changes from ConcurrentDictionay.Count to ConcurrentDictionay.Any() which is lock free but not a snapshot. 
Order of conditions reversed avoiding ConcurrentDictionay.Any() evaluation on every call.

## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)

closes #553

### CODE REVIEW

**Comment by kirankumarkolli** [comment]
V2 fix is in master is under PR review

File: Microsoft.Azure.Cosmos/src/Routing/LocationCache.cs [hunk 1]
```diff
@@ -98,8 +98,8 @@ public ReadOnlyCollection<Uri> WriteEndpoints
         {
             get
             {
-                if (this.locationUnavailablityInfoByEndpoint.Count > 0
-                    && DateTime.UtcNow - this.lastCacheUpdateTimestamp > this.unavailableLocationsExpirationTime)
+                if (DateTime.UtcNow - this.lastCacheUpdateTimestamp > this.unavailableLocationsExpirationTime
+                    && this.locationUnavailablityInfoByEndpoint.Any())
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Routing/LocationCache.cs]: Can we add a dev note explaining why so no one causes a regression in the future.

File: changelog.md [hunk 2]
```diff
@@ -1,3 +1,9 @@
+## Changes in [3.1.0](https://www.nuget.org/packages/Microsoft.Azure.Cosmos/3.0.0) : ##
+
+* Fixes
+[#558] (https://github.com/Azure/azure-cosmos-dotnet-v3/pull/558) LocationCache ConcurrentDict lock contention fix
```
Comment (simplynaveen20) [inline changelog.md]: Please remove space are [# 558] , link is coming as whole in preview

File: changelog.md [hunk 3]
```diff
@@ -1,6 +1,9 @@
 ## Changes in [3.1.0](https://www.nuget.org/packages/Microsoft.Azure.Cosmos/3.0.0) : ##
```
Comment (bartelink) [inline changelog.md]: incorrect link

Comment (kirankumarkolli) [reply changelog.md]: Thanks @bartelink . 
Using vNext as a generic representation for next release

File: changelog.md (hunk 2 above)
Comment (simplynaveen20) [reply changelog.md]: 544 is appear twice, and space is still there

Comment (kirankumarkolli) [reply changelog.md]: One space is expected. Extra space is removed.

**Pull request merged**